File: algorithms/2_medium/L/01139_largest_1_bordered_square/prefix.c
struct cell {
    int8_t i;
    int8_t j;
};

int largest1BorderedSquare(int** grid, const int rows, const int* cols_data){
    assert(rows > 0 && rows <= 100 && cols_data != NULL &&
           cols_data[0] > 0 && cols_data[0] <= 100 && grid != NULL);

    const int cols = cols_data[0];
    struct cell* prefix = calloc((size_t)(rows * cols), sizeof(*prefix));
    assert(prefix != NULL);
    for (int i = 0; i < rows; i += 1) {
        for (int j = 0; j < cols; j += 1) {
            if (grid[i][j] == 1) {
                prefix[i * cols + j].i = 1;
                prefix[i * cols + j].j = 1;
            }
            if (i > 0) {
                prefix[i * cols + j].i += prefix[(i - 1) * cols + j].i;
            }
            if (j > 0) {
                prefix[i * cols + j].j += prefix[i * cols + j - 1].j;
            }
        }
    }

    int8_t res = 0;
    for (int i = 0; i < rows; i += 1) {
        for (int j = 0; j < cols; j += 1) {
            int ni = i + res;
            int nj = j + res;
            while (ni < rows && nj < cols) {
                int8_t expect = (int8_t)(ni - i + 1);
                const int8_t up = prefix[i * cols  + nj].j - (j == 0 ? 0 : prefix[i * cols  + j - 1 ].j);
                const int8_t le = prefix[ni * cols  + j].i - (i == 0 ? 0 : prefix[(i - 1) * cols  + j].i);
                if (le != expect || up != expect) {
                    break;
                }
                const int8_t dw = prefix[ni * cols  + nj].j - (j == 0 ? 0 : prefix[ni * cols  + j - 1 ].j);
                const int8_t ri = prefix[ni * cols  + nj].i - (i == 0 ? 0 : prefix[(i - 1) * cols  + nj].i);
                if (dw == expect && ri == expect) {
                    res = expect;
                }
                ni += 1;
                nj += 1;
            }
        }
    }

/* cleanup: */
    free(prefix);
    return (int)res * res;
}

File: algorithms/2_medium/L/01139_largest_1_bordered_square/prefix.rs
impl Solution {
    pub fn largest1_bordered_square(grid: Vec<Vec<i32>>) -> i32 {
        assert!(grid.len() > 0 && grid[0].len() > 0);
        let rows = grid.len();
        let cols = grid[0].len();
        assert!(rows <= 100 && cols <= 100);
        let mut prefix = vec![(0_i8, 0_i8); rows * cols];
        for i in 0..rows {
            for j in 0..cols {
                let mut val = 0;
                if grid[i][j] == 1 {
                    prefix[i * cols + j] = (1, 1);
                }
                if i > 0 {
                    prefix[i * cols + j] .0 += prefix[(i - 1) * cols + j].0;
                }
                if j > 0 {
                    prefix[i * cols + j].1 += prefix[i * cols + j - 1].1;
                }
            }
        }

        let mut res = 0_i8;
        for i in 0..rows {
            for j in 0..cols {
                let (mut ni, mut nj) = (i + res as usize, j + res as usize);
                while ni < rows && nj < cols {
                    let expect = (ni - i + 1) as i8;
                    let up = prefix[i * cols  + nj].1 - if j == 0 { 0 } else { prefix[i * cols  + j - 1 ].1 };
                    let le = prefix[ni * cols  + j].0 - if i == 0 { 0 } else { prefix[(i - 1) * cols  + j].0 };
                    if le != expect || up != expect {
                        break;
                    }
                    let dw = prefix[ni * cols  + nj].1 - if j == 0 { 0 } else { prefix[ni * cols  + j - 1 ].1 };
                    let ri = prefix[ni * cols  + nj].0 - if i == 0 { 0 } else { prefix[(i - 1) * cols  + nj].0 };
                    if dw == expect && ri == expect {   
                        res = expect;
                    }
                    ni += 1;
                    nj += 1;
                }
            }
        }

        (res as i32).pow(2)
    }
}
